### PR TITLE
CRLP fix for null pointer exception when Membership recordtype does not exist

### LINKED ISSUE #3747: Customizable Rollups Null Pointer Exception When RecordTypeId Referenced in Filter Rule with Null Value
This surfaced in an org without a Membership Record Type. The Memberships: Won Filter Group contains a Filter Rule for Opportunity Record Type ID In List, where the value is null:

null rec type id

When a Closed Won Opportunity is inserted, there is a Null Pointer Exception.
`
Stacktrace available at the time of saving the record - Class.CMT_FilterRule.<init>: line 131, column 1 Class.CMT_FilterRuleEvaluation_SVC.evaluateFilterGroup: line 96, column 1 Class.CRLP_RollupContact_SVC.processRollups: line 83, column 1 Class.CRLP_RollupProcessor_SVC.completeRollupForParent: line 303, column 1 Class.CRLP_RollupProcessor_SVC.startRollupProcess: line 115, column 1 Class.CRLP_RollupQueueable.runRollupsForIds: line 149, column 1 Class.CRLP_RollupQueueable.execute: line 73, column 1`

The line throwing the error is in CMT_FilterRule (Line 131):
    `    } else if (filterRule.Field__r.QualifiedApiName == 'RecordTypeId' && !filterRule.Constant__c.contains(recordTypeIdPrefix)) {
`

**_Temporary Workaround if you don't have a Membership Record Type:_**
Disable all of the Membership Rollups that are associated with the Memberships: Won Filter Group.

**Steps To Repeat:**
1. Delete the Membership Opportunity Record Type.
2. Enable Customizable Rollups in NPSP Settings --> Donations --> Customizable Rollups.
3. Create a Contact (First Name = Jess; Last Name = Lopez).
4. Create an Opportunity for that Contact (Name = Test; Amount = 100; Close Date = today; Record Type = Donation, Stage = Closed Won)
5. The Opportunity saves, but the error is captured in the NPSP Errors object.

### PR DESCRIPTION
**Pull request by seanpat09**
# Critical Changes

# Changes

# Issues Closed
#3747
# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by seanpat09** [comment]
**lurch: attach W-029366

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000007fP1KQAU" target="_blank">W-029366</a>

**Comment by seanpat09** [comment]
@force2b this PR fixes the issue with the null pointer, but it does not prevent the rule from being created if the Membership record type does not exist. I tried to prevent the rule from being created if Membership record type isn't found, but I started getting this error when I turned on Customizable Rollups after deleting the Membership record type

```
Error: { "SMOE365307" : "* Rollup__mdt.NPSP_AcctLastMembershipAmount> In field: Filter_Group__c - no CustomMetadata named Filter_Group.NPSP_ClosedWon_Memberships found\n* Rollup__mdt.NPSP_AcctLastMembershipDate> In field: Filter_Group__c - no Cus" } You can try again immediately.
```

It looks like the class that builds the rollups assumes that Filter_Group.NPSP_ClosedWon_Memberships will always be created. There's more I need to untangle, but I figured this PR should go out sooner than later so that we can address the immediate bug and I'll continue to work on preventing the rule from getting created in a child branch

**Comment by force2b** [comment]
@rponti-sforg This is ready to be tested. Here are the steps I followed to configure an org that will fail with the NPE error:

**Reproduce the Error**
- New Scratch Org
- Configure NPSP Settings to exclude the "Membership" Record Type for both Donor and GAU rollups. Do not enable CRLP yet.
- On the Opportunity object, inactivate and then delete the Membership record type
- Enable CRLP
- In Setup, go to "Custom Metadata Types"
- Click the Manage link next to Filter Rules
- Find the rule for "Opps: Excl RT (HC)" (or any of the rules that say something like 'Exclude Record Types'. 
- Click the CMT name and confirm that the "Value" field on the page is empty.
- From within Setup, create a new Tab for the Error object
- Open any Account, Contact or GAU record and click the [Recalculate Rollups] button.
- Open the newly created tab for the Error object and select the 'All' view.
- There should be at least one error. Open the latest one and confirm that it shows a Null Pointer Exception error.

**Validate the Fix**
- Everything from the above
- When you get to the last step, there should be no errors at all listed.

File: src/classes/CMT_FilterRule_TEST.cls [hunk 1]
```diff
@@ -152,4 +152,103 @@ private class CMT_FilterRule_TEST {
...
+    /**
+     * @description Validates constructor can handle when Constant__c is in a date format for instead of datetime for
+     * a datetime field
+     */
+    private static testMethod void constructorShouldHandleDateTimesStoredAsDates() {
+        String filterRulesJSON =
+            CMT_UnitTestData_TEST.createFilterRuleRecord(null, null, 'TestRule1', 'Opportunity', 'CreatedDate', 'Greater', '2018-01-01');
+
+        Filter_Rule__mdt ruleMetadata = (Filter_Rule__mdt)JSON.deserialize(filterRulesJSON, Filter_Rule__mdt.class);
+        //need to null out Constant__c since null is cast as a string
```
[the remark this answers is not in the document]
Comment (seanpat09) [reply src/classes/CMT_FilterRule_TEST.cls]: that's just a copy/paste error, I'll remove the comment

File: src/classes/CRLP_Query_SEL.cls [hunk 2]
```diff
@@ -352,7 +352,7 @@
                     // If the value doesn't match though, the countOfRuleUsage map value is decremented to indicate
                     // that the Rule for the field isn't used consistently across all Filter Groups
                     if (!targetFieldAndConstant.containsKey(fieldName)) {
-                        targetFieldAndConstant.put(fieldName, rule.Constant__c);
+                        targetFieldAndConstant.put(fieldName, rule.Constant__c == null ? '' : rule.Constant__c);
```
Comment (force2b) [inline src/classes/CRLP_Query_SEL.cls]: This is a good check for null, however if an empty string is used here, then the `if` below might not match (null != ""). I suggest using something like the following above the `if` block to reference the constant value:
`String constantVal = (rule.Constant__c == null ? '' : rule.Constant__c);`

**Comment by seanpat09** [comment]
@rponti-sforg this PR is ready for testing again

**Comment by rponti-sforg** [comment]
@seanpat09 @force2b Do we need to kick off another build to see if robot tests pass?  This robot test has been failing consistently.

**Comment by davisagli** [comment]
@rponti-sforg ignore the robot build for now; I'm working on fixing a bogus failure that it's been reporting (across various branches)

**Comment by rponti-sforg** [comment]
> @rponti-sforg ignore the robot build for now; I'm working on fixing a bogus failure that it's been reporting (across various branches)

@davisagli Thanks David.  So, just wait for your fix ... then another build to kick off ... and then the Merge pull request to become green again...

**Comment by davisagli** [comment]
@rponti-sforg sorry, what I meant is that the fix is unrelated to this PR and the robot test failure should not block it from getting merged.

**Comment by davisagli** [comment]
@rponti-sforg really? We do not currently have branch protection enabled based on the Feature Test - Robot check, so that shouldn't be the case. It shows as enabled for me.

**Comment by davisagli** [comment]
@rponti-sforg That button looks enabled to me. Just because it's showing you a failed check doesn't mean that check is blocking the merge.

**Comment by seanpat09** [comment]
@rponti-sforg Just to clarify, are you referring to the `Update Branch` button that's disabled? The `Merge pull request` button below it is the one you should use.

screen shot 2018-12-19 at 8 31 32 am

**Comment by rponti-sforg** [comment]
> @rponti-sforg Just to clarify, are you referring to the `Update Branch` button that's disabled? The `Merge pull request` button below it is the one you should use.
> 
> screen shot 2018-12-19 at 8 31 32 am

@seanpat09 @davisagli Apparently I need another cup of coffee.  You are both correct ... the Merge pull request is indeed enabled.  Since I didn't see 'green' I thought there could be 'issues' if I did try and merge it...  Okay, merging ... and then getting another cup of coffee.  Thank you guys.

**Pull request merged**